sessiontxn/isolation/optimistic.go

```diff
@@ -27,28 +27,27 @@ import (
 	"go.uber.org/zap"
 )
 
+var emptyOptimisticTxnContextProvider = OptimisticTxnContextProvider{}
+
 // OptimisticTxnContextProvider provides txn context for optimistic transaction
 type OptimisticTxnContextProvider struct {
 	baseTxnContextProvider
 	optimizeWithMaxTS bool
 }
 
-// NewOptimisticTxnContextProvider returns a new OptimisticTxnContextProvider
-func NewOptimisticTxnContextProvider(sctx sessionctx.Context, causalConsistencyOnly bool) *OptimisticTxnContextProvider {
-	provider := &OptimisticTxnContextProvider{
-		baseTxnContextProvider: baseTxnContextProvider{
-			sctx:                  sctx,
-			causalConsistencyOnly: causalConsistencyOnly,
-			onTxnActive: func(_ kv.Transaction, tp sessiontxn.EnterNewTxnType) {
-				sessVars := sctx.GetSessionVars()
-				sessVars.TxnCtx.CouldRetry = isOptimisticTxnRetryable(sessVars, tp)
-			},
-		},
-	}
+// ResetForNewTxn resets OptimisticTxnContextProvider to an initial state for a new txn
+func (p *OptimisticTxnContextProvider) ResetForNewTxn(sctx sessionctx.Context, causalConsistencyOnly bool) {
+	*p = emptyOptimisticTxnContextProvider
+	p.sctx = sctx
+	p.causalConsistencyOnly = causalConsistencyOnly
+	p.onTxnActiveFunc = p.onTxnActive
+	p.getStmtReadTSFunc = p.getTxnStartTS
+	p.getStmtForUpdateTSFunc = p.getTxnStartTS
+}
```

Comment:
What about `p.optimizeWithMaxTS`?

Reply:
optimizeWithMaxTS should be false by default. `*p = emptyOptimisticTxnContextProvider` reseted all values to default.